Mathlib/Analysis/Normed/Group/Ultrametric.lean

```diff
@@ -0,0 +1,88 @@
+/-
+Copyright (c) 2024 David Loeffler. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: David Loeffler
+-/
+import Mathlib.Analysis.Normed.Group.Uniform
+import Mathlib.Topology.Algebra.Nonarchimedean.Basic
+import Mathlib.Topology.MetricSpace.Ultra.Basic
+
+/-!
+# Ultrametric normed groups
+
+We give formulations of the ultrametric triangle inequality for ultrametric spaces where the metric
+is induced by a norm; and prove that such groups are nonarchimedean groups (which is *not* true by
+definition in Mathlib – the definition of `NonarchimedeanGroup` does not refer to metrics or norms).
+-/
+
+namespace IsUltrametricDist
+
+/-- Triangle inequality for ulrametric (semi) norms on groups. -/
+@[to_additive "Triangle inequality for ulrametric (semi) norms on additive groups."]
+lemma norm_mul_le_max
+    {G : Type*} [SeminormedGroup G] [IsUltrametricDist G]
+    (x y : G) : ‖x * y‖ ≤ max ‖x‖ ‖y‖ := by
+  simpa only [dist_eq_norm_div, div_inv_eq_mul, div_one, one_mul] using dist_triangle_max x 1 y⁻¹
+
+/--
+In an commutative group with an ultrametric norm, the open balls around 1 are open subgroups.
+-/
+@[to_additive "In an commutative additive group with an ultrametric norm, the open balls around 0
+are open subgroups."]
+def ball_openSubgroup (G : Type*) [SeminormedGroup G] [IsUltrametricDist G]
+    {r : ℝ} (hr : 0 < r) : OpenSubgroup G where
+  carrier := Metric.ball (1 : G) r
+  mul_mem' {x} {y} hx hy := by
+    simp only [Metric.mem_ball, dist_eq_norm_div, div_one] at hx hy ⊢
+    exact (norm_mul_le_max x y).trans_lt (max_lt hx hy)
+  one_mem' := Metric.mem_ball_self hr
+  inv_mem' := by simp only [Metric.mem_ball, dist_one_right, norm_inv', imp_self, implies_true]
+  isOpen' := Metric.isOpen_ball
```

Comment:
```suggestion
    {r : ℝ} (hr : 0 < r) : OpenSubgroup G where
  carrier := Metric.closedBall (1 : G) r
  mul_mem' {x} {y} hx hy := by
    simp only [Metric.mem_closedBall, dist_one_right] at hx hy ⊢
    exact (norm_mul_le_max x y).trans (max_le hx hy)
  one_mem' := Metric.mem_closedBall_self hr.le
  inv_mem' := by simp
  isOpen' := IsUltrametricDist.isOpen_closedBall _ hr.ne'
```

just a thought. the proofs are just a tiny bit simpler. of course, it's morally the same.